File: shots.go
package main

import (
	"fmt"
	"math"
	"os"
	"path/filepath"
	"strconv"

	"github.com/playwright-community/playwright-go"
)

type shotsFlags struct {
	cfg           *Config
	base          string
	themes        []string
	only          []string
	scaleOverride float64
}

func runShots(f *shotsFlags) error {
	cfg := f.cfg
	SD := cfg.ShotDefaults
	OUT := abs(filepath.Join(cfg.OutDir, "shots"))
	if err := os.MkdirAll(OUT, 0o755); err != nil {
		return err
	}
	scale := SD.DeviceScaleFactor
	if f.scaleOverride > 0 {
		scale = f.scaleOverride
	}

	shots := filterByName(cfg.Shots, f.only, func(s Shot) string { return s.Name })

	fmt.Printf("shots: %s  themes=[%s]  scale=%vx\n", f.base, joinThemes(f.themes), scale)

	pw, err := playwright.Run()
	if err != nil {
		return err
	}
	defer pw.Stop()

	for _, theme := range f.themes {
		fmt.Printf("capturing shots (%s)...\n", theme)
		if err := captureShotsTheme(pw, cfg, f.base, theme, shots, SD, OUT, f.scaleOverride); err != nil {
			return err
		}
	}
	fmt.Printf("done -> %s\n", OUT)
	return nil
}

func captureShotsTheme(pw *playwright.Playwright, cfg *Config, base, theme string, shots []Shot, SD ShotDefaults, OUT string, scaleOverride float64) error {
	suf := suffix(cfg, theme)
	isDark := theme == cfg.Theme.DarkValue

	for _, shot := range shots {
		if err := captureOneShot(pw, cfg, base, theme, suf, isDark, shot, SD, OUT, scaleOverride); err != nil {
			return err
		}
	}
	return nil
}

func captureOneShot(pw *playwright.Playwright, cfg *Config, base, theme, suf string, isDark bool, shot Shot, SD ShotDefaults, OUT string, scaleOverride float64) error {
	dsf := SD.DeviceScaleFactor
	if scaleOverride > 0 {
		dsf = scaleOverride
	} else if shot.Scale != nil {
		dsf = *shot.Scale
	}

	browser, err := pw.Chromium.Launch(playwright.BrowserTypeLaunchOptions{Headless: playwright.Bool(true)})
	if err != nil {
		return err
	}
	defer browser.Close()

	colorScheme := playwright.ColorSchemeLight
	if isDark {
		colorScheme = playwright.ColorSchemeDark
	}
	ctx, err := browser.NewContext(playwright.BrowserNewContextOptions{
		Viewport:          &playwright.Size{Width: SD.Viewport.Width, Height: SD.Viewport.Height},
		DeviceScaleFactor: playwright.Float(dsf),
		ColorScheme:       colorScheme,
		ReducedMotion:     playwright.ReducedMotionReduce,
	})
	if err != nil {
		return err
	}
	if err := addThemeInitScript(ctx, cfg, theme); err != nil {
		return err
	}

	page, err := ctx.NewPage()
	if err != nil {
		return err
	}
	if shot.Live {
		if err := page.Clock().Install(playwright.ClockInstallOptions{Time: 0}); err != nil {
			return err
		}
	}
	if _, err := page.Goto(base+shot.Path, playwright.PageGotoOptions{WaitUntil: playwright.WaitUntilStateNetworkidle}); err != nil {
		return err
	}
	if _, err := page.Evaluate(`([cls, dark]) => { document.documentElement.classList.toggle(cls, dark); window.scrollTo(0, 0); }`,
		[]interface{}{cfg.Theme.DarkClass, isDark}); err != nil {
		return err
	}
	_, _ = page.Evaluate(`() => document.fonts && document.fonts.ready`)
	if shot.ForceVisible {
		applyVisibilityOverride(page)
	}

	if shot.Live && shot.AdvanceMs > 0 {
		if err := page.Clock().RunFor(shot.AdvanceMs); err != nil {
			return err
		}
		sleep(80)
	}

	out := filepath.Join(OUT, shot.Name+suf+".png")
	pad := SD.Pad
	if shot.Pad != nil {
		pad = *shot.Pad
	}

	if shot.Clip != nil {
		if _, err := page.Screenshot(playwright.PageScreenshotOptions{
			Path: playwright.String(out),
			Clip: &playwright.Rect{X: shot.Clip.X, Y: shot.Clip.Y, Width: shot.Clip.Width, Height: shot.Clip.Height},
		}); err != nil {
			return err
		}
	} else if shot.Selector != "" {
		loc := page.Locator(shot.Selector)
		if shot.Nth != nil {
			loc = loc.Nth(*shot.Nth)
		}
		if pad > 0 {
			b, err := loc.BoundingBox()
			if err != nil || b == nil {
				return fmt.Errorf("shot %q: selector not found: %s", shot.Name, shot.Selector)
			}
			if _, err := page.Screenshot(playwright.PageScreenshotOptions{
				Path: playwright.String(out),
				Clip: &playwright.Rect{
					X:      math.Max(0, b.X-float64(pad)),
					Y:      math.Max(0, b.Y-float64(pad)),
					Width:  b.Width + float64(pad*2),
					Height: b.Height + float64(pad*2),
				},
			}); err != nil {
				return err
			}
		} else {
			if _, err := loc.Screenshot(playwright.LocatorScreenshotOptions{Path: playwright.String(out)}); err != nil {
				return err
			}
		}
	} else {
		if _, err := page.Screenshot(playwright.PageScreenshotOptions{Path: playwright.String(out)}); err != nil {
			return err
		}
	}

	fmt.Printf("  %s%s.png  @%sx\n", shot.Name, suf, strconv.FormatFloat(dsf, 'f', -1, 64))
	return nil
}

File: shots.js
#!/usr/bin/env node
/*
 * High-resolution region screenshots.
 * Captures specific parts of pages (by CSS selector or explicit clip) at a high
 * deviceScaleFactor for crisp, zoomable stills. Live pages can be advanced first
 * (clock control) so canvas animations fill in before the shot.
 *
 * Usage:
 *   node shots.js                       # all shots, all themes
 *   node shots.js --theme light
 *   node shots.js --only home-headline,home-gameoflife
 *   node shots.js --scale 4             # override deviceScaleFactor
 *   node shots.js --base http://localhost:3000
 */
const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');
const cfg = require('./config.js');

const argv = process.argv.slice(2);
const opt = (n, d) => { const i = argv.indexOf(n); return i >= 0 ? argv[i + 1] : d; };
const flag = n => argv.includes(n);

const BASE = opt('--base', cfg.baseUrl);
const THEMES = opt('--theme', 'all') === 'all' ? cfg.themes : opt('--theme', 'all').split(',');
const ONLY = opt('--only', null) ? opt('--only', null).split(',') : null;
const SCALE_OVERRIDE = opt('--scale', null) ? Number(opt('--scale', null)) : null;

const SD = cfg.shotDefaults || { viewport: { width: 1600, height: 1000 }, deviceScaleFactor: 3, pad: 0 };
const OUT = path.resolve(__dirname, cfg.outDir, 'shots');
fs.mkdirSync(OUT, { recursive: true });

const sleep = ms => new Promise(r => setTimeout(r, ms));
const suffix = theme => (theme === cfg.defaultTheme ? '' : `_${theme}`);
const shotsToRun = () => (ONLY ? cfg.shots.filter(s => ONLY.includes(s.name)) : cfg.shots);

async function captureTheme(theme) {
  const suf = suffix(theme);
  const isDark = theme === cfg.theme.darkValue;
  // group by scale so we relaunch contexts only when needed
  for (const shot of shotsToRun()) {
    const dsf = SCALE_OVERRIDE || shot.scale || SD.deviceScaleFactor;
    const browser = await chromium.launch({ headless: true });
    const context = await browser.newContext({
      viewport: SD.viewport,
      deviceScaleFactor: dsf,
      colorScheme: isDark ? 'dark' : 'light',
      reducedMotion: 'no-preference',
    });
    await context.addInitScript(([k, v]) => { try { localStorage.setItem(k, v); } catch (e) {} },
      [cfg.theme.storageKey, theme]);

    const page = await context.newPage();
    if (shot.live) await page.clock.install({ time: 0 });
    await page.goto(BASE + shot.path, { waitUntil: 'networkidle' });
    await page.evaluate(([cls, dark]) => {
      document.documentElement.classList.toggle(cls, dark);
      window.scrollTo(0, 0);
    }, [cfg.theme.darkClass, isDark]);
    await page.evaluate(() => document.fonts && document.fonts.ready).catch(() => {});

    if (shot.live && shot.advanceMs) {
      await page.clock.runFor(shot.advanceMs);   // let the animation fill in
      await sleep(80);                            // let the canvas composite
    }

    const out = path.join(OUT, `${shot.name}${suf}.png`);
    const pad = shot.pad != null ? shot.pad : SD.pad;

    if (shot.clip) {
      const c = shot.clip;
      await page.screenshot({ path: out, clip: c });
    } else if (shot.selector) {
      let loc = page.locator(shot.selector);
      if (shot.nth != null) loc = loc.nth(shot.nth);
      if (pad > 0) {
        const box = await loc.boundingBox();
        await page.screenshot({ path: out, clip: {
          x: Math.max(0, box.x - pad), y: Math.max(0, box.y - pad),
          width: box.width + pad * 2, height: box.height + pad * 2,
        } });
      } else {
        await loc.screenshot({ path: out });
      }
    } else {
      await page.screenshot({ path: out }); // full viewport
    }

    const dim = await page.evaluate(() => 0); // noop
    console.log(`  ${shot.name}${suf}.png  @${dsf}x`);
    await browser.close();
  }
}

(async () => {
  console.log(`shots: ${BASE}  themes=[${THEMES}]  scale=${SCALE_OVERRIDE || SD.deviceScaleFactor}x`);
  for (const theme of THEMES) {
    console.log(`capturing shots (${theme})...`);
    await captureTheme(theme);
  }
  console.log(`done -> ${OUT}`);
})().catch(e => { console.error(e); process.exit(1); });
